data-prepper-plugins/s3-source/src/main/java/org/opensearch/dataprepper/plugins/source/ScanObjectWorker.java

```diff
@@ -0,0 +1,175 @@
+            try{
+                s3ObjectHandler.parseS3Object(s3ObjectReference,null);
+            }catch (IOException ex){
+                deleteKeyProcessedByS3Scan(s3ObjDetails);
+                LOG.error("Error while process the parseS3Object. ",ex);
+            }
+        }
+    }
+    /**
+     * Method will identify already processed key.
+     * @return boolean
+     */
+    public boolean isKeyProcessedByS3Scan(final S3ObjectDetails s3ObjectDetails) {
+        return stateSaveMap.get(s3ObjectDetails.getBucket()+s3ObjectDetails.getKey()) == null;
+    }
+
+    /**
+     * store the processed bucket key in the map.
+     */
+    public void updateKeyProcessedByS3Scan(final S3ObjectDetails s3ObjectDetails) {
+        stateSaveMap.put((s3ObjectDetails.getBucket() + s3ObjectDetails.getKey()),s3ObjectDetails);
+    }
+    private void deleteKeyProcessedByS3Scan(S3ObjectDetails s3ObjDetails) {
+        stateSaveMap.remove(s3ObjDetails.getBucket() + s3ObjDetails.getKey());
+    }
+
+    /**
+     * fetch the s3 object last modified time.
+     * @return S3ObjectDetails
+     */
+    public S3ObjectDetails getS3ObjectDetails(final S3ObjectReference s3ObjectReference){
+        GetObjectRequest getObjectRequest = GetObjectRequest.builder().bucket(s3ObjectReference.getBucketName()).key(s3ObjectReference.getKey()).build();
+        ResponseInputStream<GetObjectResponse> s3ObjectResp = s3Client.getObject(getObjectRequest);
+        final Instant instant = s3ObjectResp.response().lastModified();
+        ZonedDateTime zonedDateTime = instant.atZone(ZoneId.systemDefault());
+        return new S3ObjectDetails(s3ObjectReference.getBucketName(),s3ObjectReference.getKey(),zonedDateTime.toLocalDateTime());
+    }
+
+    /**
+     * used for identifying the slurping start date based on the range string.
+     * @return LocalDateTime
+     */
+    public LocalDateTime getSlurpingStartDateByRange(final Duration range,
+                                                     final LocalDateTime endDateTime){
+        return endDateTime.minus(range);
+    }
+
+    public LocalDateTime getSlurpingEndDateByRange(final Duration range,
+                                                     final LocalDateTime startDateTime){
+        return startDateTime.plus(range);
+    }
+
+    /**
+     * Used for identifying s3 object last modified time match with slurping date range.
+     * @return boolean
+     */
+    public boolean isKeyMatchedBetweenTimeRange(final LocalDateTime lastModifiedTime,
+                                                final Duration range,
+                                                final LocalDateTime startDateTime,
+                                                final LocalDateTime endDateTime){
+        if(Objects.nonNull(startDateTime) && Objects.nonNull(range) && Objects.nonNull(endDateTime))
+            scanRangeDateValidationError();
+        else if(Objects.nonNull(startDateTime) && Objects.nonNull(endDateTime))
```

Comment:
We can simplify this method. startDateTime, endDateTime and range are read from scanOptions and they don't change. If range is provided, we can either compute startDateTime or endDateTime based on which ever is null. This can be done in ScanObject builder.


```
// In ScanObject builder
LocalDateTime useStartDateTime;
LocalDateTime useEndDateTime;

if(Objects.nonNull(startDateTime) && Objects.nonNull(range) && Objects.nonNull(endDateTime))
            scanRangeDateValidationError();

if (Objects.isNull(startDateTime)) {
   useStartDateTime = endDateTime.minus(range);
   useEndDateTime = endDateTime;
} else if (Objects.isNull(endDateTime)) { 
  useStartDateTime = startDateTime;
  useEndDateTime = startDateTime.plus(range);
} else {
  useStartDateTime = startDateTime;
  useEndDateTime = endDateTime
}
```


```
// isKeyMatchedBetweenTimeRange can be simplified to
return lastModifiedTime.isAfter(scanOption.getUseStartDateTime())
                    && lastModifiedTime.isBefore(scanOption.getUseEndDateTime());
```